core/headerchain.go

```diff
@@ -591,353 +486,124 @@ func (hc *HeaderChain) CurrentHeader() *types.Header {
-	}
+// GetGasUsedInChain retrieves all the gas used from a given block backwards until
+// a specific distance is reached.
+func (hc *HeaderChain) CalculateBaseFee(header *types.Header) *big.Int {
+	return misc.CalcBaseFee(hc.Config(), header, hc.GetHeaderByNumber, hc.GetUnclesInChain, hc.GetGasUsedInChain)
+}
 
-	return PTP.Hash(), nil
+// Export writes the active chain to the given writer.
+func (hc *HeaderChain) Export(w io.Writer) error {
+	return hc.ExportN(w, uint64(0), hc.CurrentHeader().Number64())
 }
 
-// calcHLCRNetDifficulty calculates the net difficulty from previous prime.
-// The netDifficulties parameter inputs the nets of instantaneous difficulties from the terminus block.
-// By correctly summing the net difficulties we have obtained the proper array to be compared in HLCR.
-func (hc *HeaderChain) CalcHLCRNetDifficulty(terminalHash common.Hash, header *types.Header) ([]*big.Int, error) {
+// ExportN writes a subset of the active chain to the given writer.
+func (hc *HeaderChain) ExportN(w io.Writer, first uint64, last uint64) error {
+	hc.headermu.RLock()
+	defer hc.headermu.RUnlock()
 
-	if (terminalHash == common.Hash{}) {
-		return nil, errors.New("one or many of the  terminal hashes were nil")
+	if first > last {
+		return fmt.Errorf("export failed: first (%d) is greater than last (%d)", first, last)
 	}
+	log.Info("Exporting batch of blocks", "count", last-first+1)
 
-	primeNd := big.NewInt(0)
-	regionNd := big.NewInt(0)
-	zoneNd := big.NewInt(0)
-
-	for {
-		order, err := hc.engine.GetDifficultyOrder(header)
-		if err != nil {
-			return nil, err
-		}
-		nD := header.Difficulty[order]
-		if order <= params.PRIME {
-			primeNd.Add(primeNd, nD)
-		}
-		if order <= params.REGION {
-			regionNd.Add(regionNd, nD)
+	start, reported := time.Now(), time.Now()
+	for nr := first; nr <= last; nr++ {
+		block := hc.GetBlockByNumber(nr)
+		if block == nil {
+			return fmt.Errorf("export failed on #%d: not found", nr)
 		}
-		if order <= params.ZONE {
-			zoneNd.Add(zoneNd, nD)
+		if err := block.EncodeRLP(w); err != nil {
+			return err
 		}
-
-		if header.Hash() == terminalHash {
-			break
+		if time.Since(reported) >= statsReportLimit {
+			log.Info("Exporting blocks", "exported", block.NumberU64()-first, "elapsed", common.PrettyDuration(time.Since(start)))
+			reported = time.Now()
 		}
-
-		// Get previous header on local chain by hash
-		prevHeader := hc.GetHeaderByHash(header.ParentHash[order])
-		if prevHeader == nil {
-			// Get previous header on external chain by hash
-			prevExtBlock, err := hc.GetExternalBlock(header.ParentHash[order], header.Location, uint64(order))
-			if err != nil {
-				return nil, err
-			}
-			// Increment previous header
-			prevHeader = prevExtBlock.Header()
-		}
-		header = prevHeader
 	}
-
-	return []*big.Int{primeNd, regionNd, zoneNd}, nil
-}
-
-// SetGenesis sets a new genesis block header for the chain
-func (hc *HeaderChain) SetGenesis(head *types.Header) {
```

Comment:
Please update placement of method such that it does not look like an addition / removal.